src/IncludeInterceptor.php

```diff
@@ -76,6 +85,14 @@ public static function intercept($file, $with): void
         }
         self::$intercept = $file;
         self::$replacement = $with;
+
+        register_shutdown_function($shutdownHandler ?: static function (): void {
+            if (!self::$wasIntercepted) {
+                throw new LogicException(
+                    'Infection\'s IncludeInterceptor was not executed. Make sure you don\'t use any `file://` Stream Wrappers (like dg/bypass-finals)'
+                );
+            }
+        });
```

Comment:
1. How about a dedicated function that we will call from our side? Like `::setup(?callable $shutdownHandler)`
2. Isn't there a problem if a user happen to call `::intercept()` multiple times in a row?

Reply:
>     1. How about a dedicated function that we will call from our side? Like `::setup(?callable $shutdownHandler)`

do you mean *in addition to* current code, or instead? Could you please explain how we will use it?

What I wanted to achieve is

1. to not modify any code where `Interceptor::intercept()` is used
2. to make the default shutdown handler
3. to make it possible to override default shutdown handler to unit test the class. With the default shutdown handler - it's impossible to run unit tests without exit code equal to 0 - it always fails

So, example with `::setup()` will be useful for understanding.


> Isn't there a problem if a user happen to call ::intercept() multiple times in a row?

Nothing bad. The first handler is executed, throws an exception, and the second one is not executed. So I would say in the current state it's safe to call `::intercept()` 2 times. And I guess this will never be the case :)

But if you think we need to avoid it, I can wrap `register_shutdown_handler()` with `if` and do it only once regardless of how many times `::intercept()` is called

Reply:
ping @sanmai

Reply:
Right. I think best way to tackle the issue would be to avoid callbacks and `register_shutdown_handler()` altogether. IncludeInterceptor's job is to make intercept happen. It is not their job to control for this. That's one argument. Another reason to avoid `register_shutdown_handler()` is that it is difficult to catch exceptions thrown from there.

So I think `$wasIntercepted` is fine and good, only missing here is an accessor to read it. If we'd have a method, we can [call it, say, from here](https://github.com/infection/infection/blob/ba19fc7828f859712f8218aa573e5eb10cc78d32/src/TestFramework/Config/MutationConfigBuilder.php#L82-L83), and then we'll be free to do whatever we want, be it an exception, or a different exit code.

Reply:
> and then we'll be free to do whatever we want, be it an exception, or a different exit code.

but where we can do it? Mentions code becomes an `autoloader.php` for test framework. And this is the only one place where Infection has custom code for executed test suite.

And `register_shutdown_handler()` was a workaround to add "after test framework execution execution" event and do some checks there.

So, without using `register_shutdown_handler()`, how would you accomplish it?

Reply:
Al very least we can use this `register_shutdown_handler()` inside mentioned `autoloader.php`, but let me see if there's a better way.